Annotation scan ear libs

benjamin-confino opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Adds a system property gated routine to add ear jars to a war when annotation scanning.

fixes #30476

benjamin-confino committed:
Create test for annotation scanning in jar

benjamin-confino committed:
add ear libs to annotation scan

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=a0be97be-0770-4891-aa1f-1c3809469f46) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

benjamin-confino committed:
Better way of getting ear libs and no need to cast to web fragment

benjamin-confino committed:
better unique paths

benjamin-confino committed:
filter out null containers

benjamin-confino committed:
better logging for shared libs

benjamin-confino committed:
expand test to cover shared libs

benjamin-confino committed:
scan for manifest jars

benjamin-confino committed:
extend test to cover manifest jars

benjamin-confino committed:
all three library types are scanned, refactored for performence

benjamin-confino committed:
Add Shared Library to Test

benjamin-confino committed:
Put spring jars as shared libs so the test shared lib can load WebApplicationInitializer

benjamin-confino committed:
use PhysicalPath because shared libs don't have a nice Entry

benjamin-confino committed:
use server.xml for configuring extra lib scanning

benjamin-confino commented on dev/com.ibm.ws.container.service/bnd.bnd:
```diff
@@ -126,6 +126,10 @@ Service-Component: \
       greedy:=version;\
       configuration-policy:=optional; \
       properties:="service.vendor=IBM,toType=com.ibm.ws.container.service.config.WebFragmentsInfo", \
+   com.ibm.ws.container.service.config.WebFragmentInfoAdapter; \
```
TODO, check if this is still needed.

benjamin-confino committed:
rename applicationLibrary to earLibrary

benjamin-confino committed:
move server.xml property to enterprise application

tbitonti committed:
Prototype

benjamin-confino committed:
reregister test after git merge

tbitonti committed:
Code review corrections.

benjamin-confino committed:
correct spelling

tbitonti committed:
Fix adapt/cache calls.

benjamin-confino committed:
quick fixes, tests passing again

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=4e33d392-8adf-4b2b-9633-291db2fc734d) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

benjamin-confino committed:
Fixes for feature checker errors

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=e944b8c1-c578-46de-8e25-b4d86f12b8fa) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=78f02bf5-67e5-4e7d-b20d-0fb19cfef9cb) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=f6a3f406-f9cc-4572-81c5-bca638859df5) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.anno_fat/.classpath:
```diff
@@ -12,7 +12,12 @@
+	<classpathentry kind="src" path="test-applications/SpringAnnoTest/SpringAnnoTestManifestLib.jar/src"/>
 	<classpathentry kind="con" path="aQute.bnd.classpath.container"/>
 	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.8"/>
+	<classpathentry kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
```
Please don't add this container to the .classpath.
```suggestion
```

jhanders34 commented on dev/com.ibm.ws.anno_fat/fat/src/com/ibm/ws/tests/anno/caching/AnnotationScanInJarTest.java:
```diff
@@ -0,0 +1,121 @@
+	public static LibertyServer server;
+
+	@ClassRule
+	public static RepeatTests r = EERepeatActions.repeat(SERVER_NAME,
```
```suggestion
	public static RepeatTests r = EERepeatActions.repeat(SERVER_NAME, true,
```

No need to transform if it is already an EE 10 application.

jhanders34 commented on dev/com.ibm.ws.anno_fat/build.gradle:
```diff
@@ -23,4 +26,23 @@ dependencies {
+
+addRequiredLibraries {
+  dependsOn downloadSpringLibs	
+  dependsOn addJakartaTransformer
```
```suggestion
```

Not needed if you skip transformation in the EERepeatActions call.

jhanders34 commented on dev/com.ibm.ws.anno_fat/fat/src/com/ibm/ws/tests/anno/caching/AnnotationScanInJarTest.java:
```diff
@@ -0,0 +1,121 @@
+package com.ibm.ws.tests.anno.caching;
```
Missing copyright header in this file and the other new files added in this PR.

jhanders34 commented on dev/com.ibm.ws.app.manager.war/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -37,3 +37,6 @@ earapp.addWebModuleClassPathTo.name=Web module class path loader
 earapp.addWebModuleClassPathTo.earLoader=Use the parent enterprise application class loader for the JARs that are referenced by a web application Class-Path header.
+
+earapp.annotationScanLibaray.name=Annotation scan libraries
+earapp.annotationScanLibaray.desc=Configure which libraries are going to be scanned for Web Application annotations in a comma seperated list. Valid options are earLibrary, manifestClassPath, or ALL.
```
Should `ALL` be all caps or just `all`?

jhanders34 replied:
`Library` is spelled wrong here as `Libaray` and in the metatype.xml file as well.

jhanders34 commented on dev/com.ibm.ws.app.manager.war/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -37,8 +37,9 @@
             <Option label="%earapp.addWebModuleClassPathTo.warLoader" value="war"/>
             <Option label="%earapp.addWebModuleClassPathTo.earLoader" value="ear"/>
         </AD>
+        <AD id="annotationScanLibaray" name="%earapp.annotationScanLibaray.name" description="%earapp.annotationScanLibaray.desc" type="String" required="false"/>
```
Should this be marked beta or internal?

jhanders34 commented on dev/com.ibm.ws.app.manager/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -1,5 +1,5 @@
 ###############################################################################
-# Copyright (c) 2011, 2020 IBM Corporation and others.
+# Copyright (c) 2011, 2025 IBM Corporation and others.
```
Why updating the copyright if nothing changed?

jhanders34 commented on dev/com.ibm.ws.app.manager/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
-    Copyright (c) 2011-2020 IBM Corporation and others.
+    Copyright (c) 2011-2025 IBM Corporation and others.
```
Why updating the copyright if nothing changed?

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/ApplicationManager.java:
```diff
@@ -43,6 +43,7 @@ public class ApplicationManager {
     private long startTimeout;
     private long stopTimeout;
     private String expandLocation;
+    private String annotationScanLibaray;
```
`Library` is spelled as `Libaray` here and in other places in this class.

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/AppManagerConstants.java:
```diff
@@ -31,6 +31,7 @@ public interface AppManagerConstants {
     public static final String SHARED_APPS_DIR = WsLocationConstants.SYMBOL_SHARED_APPS_DIR;
     public static final String AUTO_INSTALL_PROP = ".installedByDropins";
     public static final String USE_JANDEX = "useJandex";
+    public static final String ANNOTATION_SCAN_LIBARAY = "annotationScanLibaray";
```
Also spelled wrong here.

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/ApplicationConfig.java:
```diff
@@ -200,6 +200,20 @@ public boolean getUseJandex() {
         return _applicationManager.getUseJandex();
     }
 
+    public String getAnnotationScanLibaray() {
```
And spelled wrong here.

jhanders34 commented on dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsAdapterImpl.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2018, 2023 IBM Corporation and others.
+ * Copyright (c) 2018, 2025 IBM Corporation and others.
```
Why updating the copyright if nothing changed?

jhanders34 commented on dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/ModuleAnnotationsImpl.java:
```diff
@@ -12,6 +12,10 @@
  *******************************************************************************/
 package com.ibm.ws.container.service.annocache.internal;
 
+import java.util.Collections;
```
Copyright date update.

jhanders34 commented on dev/com.ibm.ws.container.service/src/com/ibm/wsspi/application/handler/ApplicationInformation.java:
```diff
@@ -0,0 +1,101 @@
+/*******************************************************************************
+ * Copyright (c) 2012, 2025 IBM Corporation and others.
```
Not sure why a new class would have a starting date of 2012.  What is the value of having this added to the SPI.  How does a user get an instance of this interface? Can it be moved to a private package instead?

jhanders34 replied:
If you plan to keep this class, the package-info for this package should be updated to have the second digit updated.

jhanders34 commented on dev/com.ibm.ws.container.service/src/com/ibm/wsspi/application/handler/ApplicationInformation.java:
```diff
@@ -0,0 +1,101 @@
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
```
```suggestion
```

New classes do not need this to be part of the copyright.

benjamin-confino committed:
remove wrongly added class

benjamin-confino committed:
code review cleanup suggestions

benjamin-confino committed:
add copyright statements

benjamin-confino committed:
metatype update

benjamin-confino committed:
beta guard

benjamin-confino committed:
a few more spelling of library

benjamin-confino committed:
remove unneded entry

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=da41226e-f331-4458-ae95-8b308ebf0b9a) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=bd048bca-b9f2-4bbf-8c25-650d84d36072) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.app.manager.war/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -37,8 +37,9 @@
             <Option label="%earapp.addWebModuleClassPathTo.warLoader" value="war"/>
             <Option label="%earapp.addWebModuleClassPathTo.earLoader" value="ear"/>
         </AD>
+        <AD id="annotationScanLibrary" ibm:beta="true" name="%earapp.annotationScanLibrary.name" description="%earapp.annotationScanLibrary.desc" type="String" required="false"/>
```
Should this attribute be like the one above it where there are `<Option/>`'s with labels so that you aren't stating the valid options in the description text of the attribute because that will be self-documented in the options?  This also then explains what each option means since it will have a label.

jhanders34 replied:
No because it is a comma separated list and not just one option.

jhanders34 commented on dev/com.ibm.ws.app.manager/resources/OSGI-INF/metatype/metatype.xml:
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <!--
-    Copyright (c) 2011-2020 IBM Corporation and others.
+    Copyright (c) 2011 IBM Corporation and others.
```
```suggestion
    Copyright (c) 2011-2020 IBM Corporation and others.
```

You shouldn't remove an ending date from a copyright.  If anything it was supposed to be put back to `2011-2020` or if you are going to change it to change it to `2011, 2020`

jhanders34 commented on dev/com.ibm.ws.app.manager/resources/OSGI-INF/l10n/metatype.properties:
```diff
@@ -1,5 +1,5 @@
 ###############################################################################
-# Copyright (c) 2011, 2020 IBM Corporation and others.
+# Copyright (c) 2011 IBM Corporation and others.
```
```suggestion
# Copyright (c) 2011, 2020 IBM Corporation and others.
```

Again shouldn't remove `, 2020` from the copyright date.

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/ApplicationManager.java:
```diff
@@ -260,6 +264,33 @@ public String getExpandLocation() {
         return this.expandLocation;
     }
 
+//    private void setAnnotationScanLibrary(String annotationScanLibraryValue) {
+//
+//        annotationScanLibrary.clear();
+//        String[] split = annotationScanLibraryValue.split(",");
+//
+//        for (String s : split) {
+//            if (s.trim().equals("ALL")) {
+//                annotationScanLibrary.clear();
+//                annotationScanLibrary.add(AnnotationScanLibraryValues.ALL);
+//                break;
+//            } else if (s.trim().equals("earLib")) {
+//                annotationScanLibrary.add(AnnotationScanLibraryValues.earLib);
+//            } else if (s.trim().equals("earLib")) {
+//                annotationScanLibrary.add(AnnotationScanLibraryValues.ManifestLib);
+//            }
+//        }
+//    }
```
Why adding commented out code to the file?

benjamin-confino replied:
Tom and I made progress on enabling this for shared libraries, then hit problems and decided to focus on the easier stuff that would satisfy the customer first. We kept this so we don't loose the progress towards shared libraries.

jhanders34 replied:
Should it move then to the other class since annotationScanLibrary doesn't exists in this class any longer?

benjamin-confino replied:
You know, I think I was wrong last time and this should be removed. We are preserving stuff but looking again. This bit is both basic and obsolete

jhanders34 commented on dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/AnnotationsAdapterImpl.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2018, 2023 IBM Corporation and others.
+ * Copyright (c) 2018 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2018, 2023 IBM Corporation and others.
```

Again I don't know why removing the second date from the copyright.

benjamin-confino committed:
undo removal of second copyright and add missing copyright

jhanders34 commented on dev/com.ibm.ws.anno_fat/build.gradle:
```diff
@@ -23,4 +26,22 @@ dependencies {
+	'org.springframework:spring-core:6.1.14',
+	'org.springframework:spring-expression:6.1.14',
+	'org.springframework:spring-web:6.1.14',
+	'org.springframework:spring-webmvc:6.1.14',
+	 project(':io.openliberty.org.apache.commons.logging')
```
```suggestion
	'org.springframework:spring-webmvc:6.1.14'
```
Already part of jar in fattest.simplicity that are on the classpath

jhanders34 commented on dev/com.ibm.ws.container.service/src/com/ibm/ws/container/service/annocache/internal/WebEJBAnnotationsImpl.java:
```diff
@@ -0,0 +1,148 @@
+package com.ibm.ws.container.service.annocache.internal;
```
```suggestion
/*******************************************************************************
 * Copyright (c) 2025 IBM Corporation and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License 2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/
package com.ibm.ws.container.service.annocache.internal;
```

Copyright missing

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/wsspi/application/handler/ApplicationInformation.java:
```diff
@@ -94,4 +94,9 @@ public interface ApplicationInformation<T> {
      * @return
      */
     public boolean getUseJandex();
+
+    /**
+     * This indicates which libraries that are not usually considered war libs should be scanned for web application annotations
+     */
+    public String getAnnotationScanLibrary();
```
```suggestion
    @Deprecated
    public String getAnnotationScanLibrary();
```
I believe new APIs are beta are supposed to start out being marked `@Deprecated` until they are moved to GA.

jhanders34 replied:
```suggestion
```

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/ApplicationInstallInfo.java:
```diff
@@ -81,6 +84,12 @@ public boolean getUseJandex() {
 
+    /** {@inheritDoc} */
+    @Override
+    public String getAnnotationScanLibrary() {
```
Since this is the new SPI, there should be a beta check done here if it is called by the user.

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/ApplicationManager.java:
```diff
@@ -137,6 +138,9 @@ protected void modified(ComponentContext compcontext, Map<String, Object> proper
         setUseJandex(useJandexValue == null ? false : useJandexValue);
         //System.setProperty("com.ibm.ws.jandex.enable", useJandexValue.toString()); // Temporary -- REMOVE THIS LATER ////
 
+        String annotationScanLibraryValue = getProperty(properties, "annotationScanLibrary", "");
+        setAnnotationScanLibrary(annotationScanLibraryValue);
```
Also l believe this should be surrounded by a beta guard to not be executed unless you are using the beta.

jhanders34 commented on dev/com.ibm.ws.anno_fat/fat/src/com/ibm/ws/tests/anno/caching/AnnotationScanInJarTest.java:
```diff
@@ -0,0 +1,130 @@
+
+
+	@Test
+	public void testSpringAnnotationFoundInwar() throws Exception {
```
This is the only test and it only tests the `all` option.  Seems you should update com.ibm.websphere.simplicity.config in fattest.simplicity to include the new option there and have additional tests that test the other two options to make sure that the right output comes out for those as well.

benjamin-confino committed:
beta fencing

benjamin-confino committed:
add enterpirse application config class

benjamin-confino committed:
split test into one for each config option

benjamin-confino committed:
correct config keywords

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=83631258-791d-44bf-b333-28f4201cd2e7) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.anno_fat/fat/src/com/ibm/ws/tests/anno/caching/extrascan/AbstractAnnotationScanInJarTest.java:
```diff
@@ -0,0 +1,164 @@
+		JavaArchive sharedLib = ShrinkWrap.create(JavaArchive.class, "sharedLib.jar")
+				.addPackages(true, SharedLibInit.class.getPackage());
+
+		JavaArchive maifestJar = ShrinkWrap.create(JavaArchive.class, "manifestLib.jar")
+				.addPackage(ManifestInit.class.getPackage());
+
+		EnterpriseArchive ear = ShrinkWrap.create(EnterpriseArchive.class, APP_NAME + ".ear")
+				.addAsModule(war)
+				.addAsModule(maifestJar)
```
It is a small nit, but I figure you want the variable name to be `manifestJar`

```suggestion
		JavaArchive manifestJar = ShrinkWrap.create(JavaArchive.class, "manifestLib.jar")
				.addPackage(ManifestInit.class.getPackage());

		EnterpriseArchive ear = ShrinkWrap.create(EnterpriseArchive.class, APP_NAME + ".ear")
				.addAsModule(war)
				.addAsModule(manifestJar)
```

jhanders34 commented on dev/com.ibm.ws.app.manager/bnd.bnd:
```diff
@@ -59,7 +59,9 @@ instrument.classesExcludes: com/ibm/ws/app/manager/internal/resources/*.class
 	com.ibm.ws.kernel.boot.common;version=latest,\
-	com.ibm.ws.kernel.feature.core;version=latest
+	com.ibm.ws.kernel.feature.core;version=latest,\
+	com.ibm.ws.kernel.boot.core,\
+	com.ibm.ws.kernel.boot.core
```
```suggestion
	com.ibm.ws.kernel.boot.core
```
Not reason to have it in the build path twice.

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/ApplicationManager.java:
```diff
@@ -137,6 +143,13 @@ protected void modified(ComponentContext compcontext, Map<String, Object> proper
+            String annotationScanLibraryValue = getProperty(properties, "annotationScanLibrary", "");
+            setAnnotationScanLibrary(annotationScanLibraryValue);
+        } else {
+            setAnnotationScanLibrary("");
```
Should we call it at all if it isn't beta.

benjamin-confino replied:
Leaving it as null will cause an NPE down the line, but setting it to blank will ensure the new routines are cut down to if statements that get that empty string and return false. I could betafence those if statements too.

jhanders34 commented on dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/EnterpriseApplication.java:
```diff
@@ -0,0 +1,69 @@
+import javax.xml.bind.annotation.XmlAttribute;
+
+/**
+ * Represents a &lt;webApplication> configuration element
```
```suggestion
 * Represents a &lt;enterpriseApplication> configuration element
```

jhanders34 replied:
It is still unchanged

benjamin-confino replied:
Oops, fixed.

jhanders34 commented on dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/EnterpriseApplication.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public String toString() {
+        StringBuffer buf = new StringBuffer("WebApplication{");
```
```suggestion
        StringBuilder buf = new StringBuilder("EnterpriseApplication{");
```

jhanders34 commented on dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/EnterpriseApplication.java:
```diff
@@ -0,0 +1,69 @@
+ */
+public class EnterpriseApplication extends Application {
+
+    private String contextRoot;
+    private String annotationScanLibrary;
+
+    /**
+     * @return the contextRoot
+     */
+    public String getContextRoot() {
+        return contextRoot;
+    }
+
+    /**
+     * @param contextRoot the contextRoot to set
+     */
+    @XmlAttribute
+    public void setContextRoot(String contextRoot) {
+        this.contextRoot = contextRoot;
+    }
```
```suggestion
    private String annotationScanLibrary;
```
contextRoot is not applicable to EnterpriseApplication from what I am seeing.  I believe you probably should add other attributes that are valid though like `defaultClientModule` and `webModuleClassPathLoader`.

benjamin-confino replied:
I was going for a minimum viable product, but sure, why not do the maximal impl. Its not that much larger. 

But are you sure about context root? In I see com.ibm.ws.app.manager.war/resources/OSGI-INF/metatype/metatype.xml

```    
<OCD description="%earapp.desc" name="%earapp.name" id="com.ibm.ws.app.manager.earappcfg" ibm:alias="enterpriseApplication" ibm:extends="com.ibm.ws.app.manager">
        <AD id="type" name="internal" description="internal" type="String" default="ear" ibm:final="true"/>
        <AD id="context-root" name="internal" description="internal" ibm:final="true" required="false" type="String"/> 
```
        
Ooh, its a context-root and webapp is contextRoot. That's devious.

jhanders34 replied:
And context-root is internal.  It is not an external attribute to be set in server.xml.

benjamin-confino replied:
Got you, in that case I'll remove it.

jhanders34 commented on dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/EnterpriseApplication.java:
```diff
@@ -0,0 +1,69 @@
+    public String toString() {
+        StringBuffer buf = new StringBuffer("WebApplication{");
+        buf.append(super.toString());
+        if (contextRoot != null)
+            buf.append("contextRoot=\"" + contextRoot + "\" ");
```
This would go away with my previous review comment, and `annotationScanLibrary` and the additional attributes if you include them would go here as well.

jhanders34 commented on dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/ServerConfiguration.java:
```diff
@@ -906,6 +909,16 @@ public ConfigElementList<WebApplication> getWebApplications() {
     }
 
+    /**
+     * @return explicitly installed web applications
```
```suggestion
     * @return explicitly installed enterprise applications
```

benjamin-confino committed:
expand EnterpriseApplication to complete impl

benjamin-confino committed:
correct spelling

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=daf3884e-6fa0-4af7-8a2b-d0407caf8fca) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

benjamin-confino committed:
replace prohibited term

benjamin-confino committed:
Remove context-root attribute

benjamin-confino committed:
Revert "beta guard"

This reverts commit dbb8d7e820943c5e2793d6375cc9d2aa990461e3.

benjamin-confino committed:
Revert "beta fencing"

This reverts commit 1b962443ac765afd8dc4e38d129f6aa6338dd005.

benjamin-confino commented:
!build  ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=aa14bb1a-2260-4b07-9596-e0fe21242fd2) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.anno_fat/fat/src/com/ibm/ws/tests/anno/FATSuite.java:
```diff
@@ -67,8 +71,12 @@
     FragmentOrderTest.class,
     LooseConfigTest.class,
-    AppReplaceTest.class
+
```
```suggestion
    AppReplaceTest.class
```

On merge, you removed this test from the list and it shouldn't be removed.

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/ApplicationManager.java:
```diff
@@ -137,6 +138,9 @@ protected void modified(ComponentContext compcontext, Map<String, Object> proper
         setUseJandex(useJandexValue == null ? false : useJandexValue);
         //System.setProperty("com.ibm.ws.jandex.enable", useJandexValue.toString()); // Temporary -- REMOVE THIS LATER ////
 
+        String annotationScanLibraryValue = getProperty(properties, "annotationScanLibrary", "");
```
Shouldn't the changes in this entire class be undone?  `com.ibm.ws.app.manager/resources/OSGI-INF/metatype/metatype.xml` wasn't updated to add an `annotationScanLibrary` attribute to app manager config so this will always be null converted to `""`.

benjamin-confino replied:
Perhaps. If I recall I followed the call chain which led me to adding stuff like this. I'll undo it and test it. If it works without it then I'll commit the revert.

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/ApplicationConfig.java:
```diff
@@ -200,6 +200,20 @@ public boolean getUseJandex() {
         return _applicationManager.getUseJandex();
     }
 
+    public String getAnnotationScanLibrary() {
```
Should this logic move to the `EARDeployedAppInfo` class and be similar to how we handle `webModuleClassPathLoader`?  Similarly the `ANNOTATION_SCAN_LIBRARY` constants comes out of ApplicationManagerConstants and moves to `EarDeployedAppInfo`

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/ApplicationInstallInfo.java:
```diff
@@ -81,6 +84,12 @@ public boolean getUseJandex() {
     }
 
+    /** {@inheritDoc} */
+    @Override
```
I think this logic then goes away as well?

jhanders34 commented on dev/com.ibm.ws.kernel.metatype.helper/src/com/ibm/wsspi/config/internal/FilesetImpl.java:
```diff
@@ -368,6 +368,35 @@ private void getHumanReadableSummary(StringBuilder sb) {
         }
 
+       public boolean hasFilters() {
+            return !includes.isEmpty() && excludes.isEmpty();
```
```suggestion
            return !includes.isEmpty() && !excludes.isEmpty();
```
I assume this was wrong.

jhanders34 commented on dev/com.ibm.ws.kernel.metatype.helper/src/com/ibm/wsspi/config/internal/FilesetImpl.java:
```diff
@@ -368,6 +368,35 @@ private void getHumanReadableSummary(StringBuilder sb) {
+
+        public String getHumanReadableSummary() {
+            StringBuilder sb = new StringBuilder();
+
```
This method is odd.  Why not just:
```
            getHumanReadableSummary(sb);
            return sb.toString();
```

benjamin-confino committed:
Code review update
Add test that was removed in a merge
Remove attempt to read the new property from ApplicationManager which will always return and empty string
Do not attempt to use the application manager in ApplicationConfig
Remove unused methods in FilesetImpl

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/ApplicationInstallInfo.java:
```diff
@@ -81,6 +84,12 @@ public boolean getUseJandex() {
+    /** {@inheritDoc} */
+    @Override
+    public String getAnnotationScanLibrary() {
+        return _config.getAnnotationScanLibrary();
```
```suggestion
        return (String) _config.getConfigProperty(AppManagerConstants.ANNOTATION_SCAN_LIBRARY);
```
Of course you will need to add the import for AppManagerConstants as well.

jhanders34 replied:
```suggestion
        Object result = _config.getConfigProperty(AppManagerConstants.ANNOTATION_SCAN_LIBRARY);
        if (result instanceof String) {
            return (String) result;
        }

        // If we haven't found it return an empty string
        return "";
```

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/ApplicationConfig.java:
```diff
@@ -200,6 +200,20 @@ public boolean getUseJandex() {
         return _applicationManager.getUseJandex();
     }
 
+    public String getAnnotationScanLibrary() {
+        // First try to get the value from the application configuration
+        // which overrides the value on the application manager configuration.
+        if (_config != null) {
+            Object result = _config.get(AppManagerConstants.ANNOTATION_SCAN_LIBRARY);
+            if (result instanceof String) {
+                return (String) result;
+            }
+        }
+
+        // The application manager does not provide this property, so if we havne't found it return an empty string
+        return "";
+    }
+
```
```suggestion
```

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/ws/app/manager/internal/ApplicationInstallInfo.java:
```diff
@@ -81,6 +84,12 @@ public boolean getUseJandex() {
         return _config.getUseJandex();
     }
 
+    /** {@inheritDoc} */
+    @Override
+    public String getAnnotationScanLibrary() {
+        return _config.getAnnotationScanLibrary();
+    }
+
```
```suggestion
```

benjamin-confino committed:
Avoid need to update external SPI

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

benjamin-confino committed:
Changes to bnd and features are no longer needed with the refactoring

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=52b18f9e-1ab4-42cf-bc39-942cb87546da) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/wsspi/application/handler/package-info.java:
```diff
@@ -11,7 +11,7 @@
  *******************************************************************************/
 /**
- * @version 1.2
+ * @version 1.3
```
```suggestion
 * @version 1.2
```

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/wsspi/application/handler/package-info.java:
```diff
@@ -11,7 +11,7 @@
+ * @version 1.3
  */
-@org.osgi.annotation.versioning.Version("1.2")
+@org.osgi.annotation.versioning.Version("1.3")
```
```suggestion
@org.osgi.annotation.versioning.Version("1.2")
```

jhanders34 commented on dev/com.ibm.ws.app.manager/src/com/ibm/wsspi/application/handler/package-info.java:
```diff
@@ -1,5 +1,5 @@
 /*******************************************************************************
- * Copyright (c) 2017 IBM Corporation and others.
+ * Copyright (c) 2017, 2025 IBM Corporation and others.
```
```suggestion
 * Copyright (c) 2017 IBM Corporation and others.
```

benjamin-confino committed:
Undo copyright change as the changes were reverted

benjamin-confino committed:
Undo update to package version that is no longer needed

Co-authored-by: Jared Anderson <jhanders@us.ibm.com>

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=6f982125-f674-4537-b11a-180d16e48128) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

benjamin-confino commented:
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=ac3ee708-edc3-4b37-97c9-478edb7da6fa) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._
